tests/test_studio.py

```diff
@@ -0,0 +1,109 @@
+    mocker.patch("scmrepo.git.Git")
+    mocked_post = mocker.patch("requests.post")
+    monkeypatch.setenv(env.STUDIO_URL, "STUDIO_URL")
+    monkeypatch.setenv(env.STUDIO_TOKEN, "STUDIO_TOKEN")
+
+    live = Live()
+    mocked_post.assert_called_with(
+        "https://0.0.0.0",
+        json={
+            "type": "start",
+            "repo_url": "STUDIO_URL",
+            "rev": mocker.ANY,
+            "step": 0,
+            "timestamp": mocker.ANY,
+            "metrics": {},
+            "plots": {},
+            "client": "dvclive",
+        },
+        headers={"Authorization": "token STUDIO_TOKEN"},
+    )
+
+    live.log("foo", 1)
+
+    live.next_step()
+    mocked_post.assert_called_with(
+        "https://0.0.0.0",
+        json={
+            "type": "data",
+            "repo_url": "STUDIO_URL",
+            "rev": mocker.ANY,
+            "step": 0,
+            "timestamp": mocker.ANY,
+            "metrics": {"dvclive.json": {"data": {"step": 0, "foo": 1}}},
+            "plots": {
+                "dvclive/scalars/foo.tsv": {
+                    "data": [
+                        {"timestamp": mocker.ANY, "step": "0", "foo": "1"}
+                    ]
+                }
+            },
+            "client": "dvclive",
+        },
+        headers={"Authorization": "token STUDIO_TOKEN"},
+    )
+
+    live.end()
+    mocked_post.assert_called_with(
+        "https://0.0.0.0",
+        json={
+            "type": "completed",
+            "repo_url": "STUDIO_URL",
+            "rev": mocker.ANY,
+            "step": live.get_step(),
+            "timestamp": mocker.ANY,
+            "metrics": {},
+            "plots": {},
+            "client": "dvclive",
+        },
+        headers={"Authorization": "token STUDIO_TOKEN"},
+    )
+
+
+@pytest.mark.studio
+def test_post_to_studio_failed_request(tmp_dir, mocker, monkeypatch):
+    mocker.patch("scmrepo.git.Git")
+
+    mocker.patch("requests.post")
+    monkeypatch.setenv(env.STUDIO_URL, "STUDIO_URL")
+    monkeypatch.setenv(env.STUDIO_TOKEN, "STUDIO_TOKEN")
+
+    live = Live()
+
+    mocked_response = mocker.MagicMock()
+    mocked_response.status_code.return_value = 400
+    mocker.patch("requests.post", return_value=mocked_response)
+    live.log("foo", 1)
+    live.next_step()
+
+    mocked_post = mocker.patch("requests.post")
+    live.log("foo", 2)
+    live.next_step()
+    mocked_post.assert_called_with(
+        "https://0.0.0.0",
+        json={
+            "type": "data",
+            "repo_url": "STUDIO_URL",
+            "rev": mocker.ANY,
+            "step": 1,
+            "timestamp": mocker.ANY,
+            "metrics": {"dvclive.json": {"data": {"step": 1, "foo": 2}}},
+            "plots": {
+                "dvclive/scalars/foo.tsv": {
+                    "data": [
+                        {"timestamp": mocker.ANY, "step": "0", "foo": "1"},
+                        {"timestamp": mocker.ANY, "step": "1", "foo": "2"},
+                    ]
```

Comment:
@Suor Is this what Studio expects when a previous request has failed? I just include the datapoints that were not sent

Reply:
Is `step` guaranteed to be there? Can I rely on that if I saved some plot row with `step == X` then if I receive it again it will be the same thing resend, so that I may ignore it safely? Why are those strings? May I convert them to integer for purposes of order? Is it the same step as in root of the event - which is needed to filter out resending if metrics too?

Reply:
> Is step guaranteed to be there?

Yes

> Can I rely on that if I saved some plot row with step == X then if I receive it again it will be the same thing resent to me, so that I may ignore it safely?

Yes. But I will only send older steps if Studio doesn't return a success code.

> Why are those strings? May I convert them to integer for purposes of order?

Just that the `csv` reader returns strings by default. I can cast to integer before posting.

> Is it the same step as in root of the event - which is needed to filter out resending if metrics too?

The highest step number in `plots` will always match the `step` in the top-level request. 

I am not currently resending metrics from failed requests, just the latest ones. Should I also send metrics from previous steps?